File: firmware/test/ws/test_ws1.py
#!/usr/bin/env python

# VSCP ws1 client example  (Need python3)
# Demonstrates the use of the ws1 websocket interface of the VSCP daemon
# Sample event to send
#    send 0,20,3,,,0,-,15,14,13,12,11,10,9,8,7,6,5,4,3,2,0,0,1,35
# Original sample from: https://websockets.readthedocs.io/en/stable/intro.html
# Copyright 2020-2026 Ake Hedman, the VSCP project - MIT license

from signal import signal, SIGINT
from sys import exit
import sys
import os
import asyncio
import websockets
import re

DEFAULT_URL = 'ws://192.168.1.100:8884/ws1'
DEFAULT_USERNAME = 'vscp'
DEFAULT_PASSWORD = 'secret'
DEFAULT_KEY_HEX = 'A4A86F7D7E119BA3F0CD06881E371B98'

HEX_32_RE = re.compile(r'^[0-9a-fA-F]{32}$')
HEX_32_ANY_RE = re.compile(r'\b([0-9a-fA-F]{32})\b')


def get_aes_backend_name() -> str:
    try:
        from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
        _ = (Cipher, algorithms, modes)
        return 'cryptography'
    except ImportError:
        pass

    try:
        from Crypto.Cipher import AES
        _ = AES
        return 'pycryptodome'
    except ImportError:
        return 'missing'


def _pkcs7_pad(data: bytes, block_size: int = 16) -> bytes:
    pad_len = block_size - (len(data) % block_size)
    return data + bytes([pad_len]) * pad_len


def encrypt_auth_credentials(username: str, password: str, sid_hex: str, key_hex: str) -> str:
    if not HEX_32_RE.fullmatch(sid_hex or ''):
        raise ValueError(f"Invalid SID for IV: expected 32 hex chars, got '{sid_hex}'")

    if not HEX_32_RE.fullmatch(key_hex or ''):
        raise ValueError('Invalid AES-128 key: expected 32 hex chars')

    iv = bytes.fromhex(sid_hex)
    key = bytes.fromhex(key_hex)
    plaintext = _pkcs7_pad(f"{username}:{password}".encode('utf-8'))

    try:
        from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
        encryptor = Cipher(algorithms.AES(key), modes.CBC(iv)).encryptor()
        ciphertext = encryptor.update(plaintext) + encryptor.finalize()
        return ciphertext.hex()
    except ImportError:
        try:
            from Crypto.Cipher import AES
            cipher = AES.new(key, AES.MODE_CBC, iv)
            return cipher.encrypt(plaintext).hex()
        except ImportError as ex:
            raise RuntimeError('AES dependency missing. Install `cryptography` or `pycryptodome`.') from ex


def extract_sid_from_text(text: str):
    if not text:
        return None

    match = HEX_32_ANY_RE.search(text)
    if match:
        return match.group(1)

    return None


def get_cli_config():
    args = sys.argv[1:]

    if '--help' in args or '-h' in args:
        print('Usage: python test_ws1.py [url] [--url <url>] [--user <username>] [--password <password>] [--key <aes128-hex>]')
        print('Example: python test_ws1.py --url ws://192.168.1.50:8884/ws1 --user admin --password secret --key A4A86F7D7E119BA3F0CD06881E371B98')
        print('Env: WS1_URL, WS1_USERNAME, WS1_PASSWORD, WS1_KEY (fallback: VSCP_USERNAME, VSCP_PASSWORD, VSCP_KEY16)')
        exit(0)

    url = None
    username = None
    password = None
    key_hex = None
    positionals = []

    index = 0
    while index < len(args):
        arg = args[index]

        if arg == '--url' and index + 1 < len(args):
            url = args[index + 1]
            index += 2
            continue

        if arg.startswith('--url='):
            url = arg[len('--url='):]
            index += 1
            continue

        if arg == '--user' and index + 1 < len(args):
            username = args[index + 1]
            index += 2
            continue

        if arg.startswith('--user='):
            username = arg[len('--user='):]
            index += 1
            continue

        if arg == '--password' and index + 1 < len(args):
            password = args[index + 1]
            index += 2
            continue

        if arg.startswith('--password='):
            password = arg[len('--password='):]
            index += 1
            continue

        if arg == '--key' and index + 1 < len(args):
            key_hex = args[index + 1]
            index += 2
            continue

        if arg.startswith('--key='):
            key_hex = arg[len('--key='):]
            index += 1
            continue

        if not arg.startswith('-'):
            positionals.append(arg)

        index += 1

    if url is None and positionals:
        url = positionals[0]

    if not url:
        url = os.environ.get('WS1_URL', DEFAULT_URL)

    if not username:
        username = os.environ.get('WS1_USERNAME') or os.environ.get('VSCP_USERNAME') or DEFAULT_USERNAME

    if not password:
        password = os.environ.get('WS1_PASSWORD') or os.environ.get('VSCP_PASSWORD') or DEFAULT_PASSWORD

    if not key_hex:
        key_hex = os.environ.get('WS1_KEY') or os.environ.get('VSCP_KEY16') or DEFAULT_KEY_HEX

    return {
        'url': url,
        'username': username,
        'password': password,
        'key_hex': key_hex.strip(),
    }


def handler(signal_received, frame):
    print('SIGINT or CTRL-C detected. Exiting')
    exit(0)


async def connect(config):
    async with websockets.connect(
                 config['url'], ping_interval=None, ping_timeout=1, close_timeout=100) as websocket:

        await websocket.send("C;CHALLENGE")
        greeting = await websocket.recv()
        print(f"Response from server: {greeting}")

        sid = extract_sid_from_text(greeting)
        if not sid:
            raise RuntimeError('No SID found in challenge response')

        auth_crypto = encrypt_auth_credentials(config['username'], config['password'], sid, config['key_hex'])
        cmdauth = f"C;AUTH;{sid};{auth_crypto}"

        print(f"Logging in as {config['username']}")
        print(f"AUTH IV   : {sid}")
        print(f"AUTH CRYPT: {auth_crypto}")
        await websocket.send(cmdauth)
        greeting = await websocket.recv()
        print(f"Response from server: {greeting}")

        await websocket.send("C;OPEN")
        greeting = await websocket.recv()
        print(f"Response from server: {greeting}")

        print("Waiting for events")
        while True:
            greeting = await websocket.recv()
            print(f"Response from server: {greeting}")


if __name__ == '__main__':
    signal(SIGINT, handler)
    cli_config = get_cli_config()
    backend = get_aes_backend_name()
    if backend == 'missing':
        print('AES backend : missing (install cryptography or pycryptodome)')
    else:
        print(f'AES backend : {backend}')
    print(f"Connecting to {cli_config['url']}")
    asyncio.get_event_loop().run_until_complete(connect(cli_config))

File: firmware/test/ws/test_ws1.js
/*!
  Example code for node.js and the VSCP daemon ws1 websocket interface
  send 0,30,5,,,0,-,15,14,13,12,11,10,9,8,7,6,5,4,3,2,0,0,1,35
  Code from: https://github.com/websockets/ws

  Copyright 2020-2026 Ake Hedman, the VSCP project - MIT license
*/

WebSocket = require('ws');
const crypto = require('crypto');

const defaultUrl = 'ws://192.168.1.104:8884/ws1';
const DEFAULT_USERNAME = 'vscp';
const DEFAULT_PASSWORD = 'secret';
const DEFAULT_KEY_HEX = 'A4A86F7D7E119BA3F0CD06881E371B98';

function parseCliArgs() {
  const args = process.argv.slice(2);
  const parsed = {
    help: false,
    url: null,
    username: null,
    password: null,
    keyHex: null,
    positionals: []
  };

  for (let index = 0; index < args.length; index++) {
    const arg = args[index];

    if (arg === '--help' || arg === '-h') {
      parsed.help = true;
      continue;
    }

    if (arg === '--url' && args[index + 1]) {
      parsed.url = args[++index];
      continue;
    }

    if (arg.startsWith('--url=')) {
      parsed.url = arg.substring('--url='.length);
      continue;
    }

    if (arg === '--user' && args[index + 1]) {
      parsed.username = args[++index];
      continue;
    }

    if (arg.startsWith('--user=')) {
      parsed.username = arg.substring('--user='.length);
      continue;
    }

    if (arg === '--password' && args[index + 1]) {
      parsed.password = args[++index];
      continue;
    }

    if (arg.startsWith('--password=')) {
      parsed.password = arg.substring('--password='.length);
      continue;
    }

    if (arg === '--key' && args[index + 1]) {
      parsed.keyHex = args[++index];
      continue;
    }

    if (arg.startsWith('--key=')) {
      parsed.keyHex = arg.substring('--key='.length);
      continue;
    }

    if (!arg.startsWith('-')) {
      parsed.positionals.push(arg);
    }
  }

  return parsed;
}

function getAuthConfig() {
  const cli = parseCliArgs();

  if (cli.help) {
    console.log('Usage: node test_ws1.js [url] [--url <url>] [--user <username>] [--password <password>] [--key <aes128-hex>]');
    console.log('Example: node test_ws1.js --url ws://192.168.1.100:8884/ws1 --user admin --password secret --key A4A86F7D7E119BA3F0CD06881E371B98');
    console.log('Env: WS1_URL, WS1_USERNAME, WS1_PASSWORD, WS1_KEY (fallback: VSCP_USERNAME, VSCP_PASSWORD, VSCP_KEY16)');
    process.exit(0);
  }

  const wsUrl = cli.url || cli.positionals[0] || process.env.WS1_URL || defaultUrl;
  const username = cli.username || process.env.WS1_USERNAME || process.env.VSCP_USERNAME || DEFAULT_USERNAME;
  const password = cli.password || process.env.WS1_PASSWORD || process.env.VSCP_PASSWORD || DEFAULT_PASSWORD;
  const keyHex = (cli.keyHex || process.env.WS1_KEY || process.env.VSCP_KEY16 || DEFAULT_KEY_HEX).trim();

  return {
    wsUrl,
    username,
    password,
    keyHex
  };
}

function isValidHex(value, expectedLength) {
  return typeof value === 'string' && value.length === expectedLength && /^[0-9a-fA-F]+$/.test(value);
}

function encryptAuthCredentials(username, password, sidHex, keyHex) {
  if (!isValidHex(sidHex, 32)) {
    throw new Error(`Invalid SID for IV: expected 32 hex chars, got '${sidHex}'`);
  }

  if (!isValidHex(keyHex, 32)) {
    throw new Error('Invalid AES-128 key: expected 32 hex chars');
  }

  const iv = Buffer.from(sidHex, 'hex');
  const key = Buffer.from(keyHex, 'hex');
  const plainCredentials = Buffer.from(`${username}:${password}`, 'utf8');

  const cipher = crypto.createCipheriv('aes-128-cbc', key, iv);
  const encrypted = Buffer.concat([cipher.update(plainCredentials), cipher.final()]);
  return encrypted.toString('hex');
}

const authConfig = getAuthConfig();
const wsUrl = authConfig.wsUrl;
console.log(`Connecting to ${wsUrl}`);

const ws = new WebSocket(wsUrl, {
  perMessageDeflate: false
});

const REQUEST_TIMEOUT_MS = 5000;
let pendingRequest = null;
let pendingSidResolver = null;

function toTextMessage(payload) {
  if (typeof payload === 'string') {
    return payload;
  }

  if (Buffer.isBuffer(payload)) {
    return payload.toString('utf8');
  }

  if (payload instanceof ArrayBuffer) {
    return Buffer.from(payload).toString('utf8');
  }

  if (ArrayBuffer.isView(payload)) {
    return Buffer.from(payload.buffer, payload.byteOffset, payload.byteLength).toString('utf8');
  }

  return String(payload);
}

function parseWs1Event(eventPayload) {
  const parts = eventPayload.split(',');
  if (parts.length < 7) {
    return null;
  }

  return {
    head: parts[0],
    vscpClass: parts[1],
    vscpType: parts[2],
    obid: parts[3],
    datetime: parts[4],
    timestamp: parts[5],
    guid: parts[6],
    data: parts.slice(7)
  };
}

function parseWs1Message(message) {
  if (!message || message.length < 1) {
    return { kind: 'unknown', raw: message };
  }

  const kind = message[0];
  const fields = message.split(';');

  if (kind === '+') {
    return {
      kind: 'positive',
      command: fields[1] || '',
      details: fields.slice(2),
      raw: message
    };
  }

  if (kind === '-') {
    return {
      kind: 'negative',
      command: fields[1] || '',
      errorCode: fields[2] || '',
      errorText: fields.slice(3).join(';'),
      raw: message
    };
  }

  if (kind === 'E') {
    return {
      kind: 'event',
      event: parseWs1Event(fields.slice(1).join(';')),
      raw: message
    };
  }

  return { kind: 'unknown', raw: message };
}

function getWs1CommandFromFrame(frame) {
  const fields = frame.split(';');
  if (fields.length >= 2 && fields[0] === 'C') {
    return (fields[1] || '').toUpperCase();
  }

  return '';
}

function matchesExpectedCommand(parsed, expectedCommand) {
  if (!expectedCommand) {
    return true;
  }

  return (parsed.command || '').toUpperCase() === expectedCommand.toUpperCase();
}

function sendAndWait(commandFrame, expectedCommand) {
  if (pendingRequest) {
    return Promise.reject(new Error('A command is already waiting for response'));
  }

  const expected = expectedCommand || getWs1CommandFromFrame(commandFrame);

  return new Promise((resolve, reject) => {
    const timeoutId = setTimeout(() => {
      pendingRequest = null;
      reject(new Error(`Timeout waiting for response to ${commandFrame}`));
    }, REQUEST_TIMEOUT_MS);

    pendingRequest = {
      expected,
      resolve: (parsed) => {
        clearTimeout(timeoutId);
        pendingRequest = null;
        resolve(parsed);
      },
      reject: (error) => {
        clearTimeout(timeoutId);
        pendingRequest = null;
        reject(error);
      }
    };

    ws.send(commandFrame);
  });
}

function extractSid(parsed) {
  if (!parsed || parsed.kind !== 'positive') {
    return null;
  }

  const sidCandidate = (parsed.details || []).find((field) => /^[0-9a-fA-F]{32}$/.test(field));
  return sidCandidate || null;
}

function waitForSid() {
  if (pendingSidResolver) {
    return Promise.reject(new Error('Already waiting for SID'));
  }

  return new Promise((resolve, reject) => {
    let challengeSent = false;
    let timeoutId;

    function armTimeout() {
      timeoutId = setTimeout(() => {
        if (!challengeSent) {
          challengeSent = true;
          console.log('* * * SID wait timed out, requesting CHALLENGE...');
          ws.send('C;CHALLENGE');
          armTimeout();
          return;
        }

        pendingSidResolver = null;
        reject(new Error('Timeout waiting for SID (including CHALLENGE retry)'));
      }, REQUEST_TIMEOUT_MS);
    }

    pendingSidResolver = {
      resolve: (sid) => {
        clearTimeout(timeoutId);
        pendingSidResolver = null;
        resolve(sid);
      },
      reject: (error) => {
        clearTimeout(timeoutId);
        pendingSidResolver = null;
        reject(error);
      }
    };

    armTimeout();
  });
}

ws.on('open', async function open() {
  const cmdnoop = "C;NOOP";
  const cmdopen = "C;OPEN";

  // 'E';head,vscp_class,vscp_type,obid,datetime,timestamp,GUID,data
  const sendEvent = "E;0,30,5,0,2020-01-29T23:05:59Z,0,FF:FF:FF:FF:FF:FF:FF:F5:00:00:00:00:00:02:00:00,1,2,3,4,5,6";

  // Receive only CLASS1.CONTROL, TurnOn

  // C;SF;filter-priority, filter-class, 
  //  filter-type, filter-GUID;mask-priority, 
  //  mask-class, mask-type, mask-GUID”
  const cmdFilter = "C;SF;0,30,5,00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00;0,0xffff,0xffff,00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00";

  try {
    console.log("\n* * * Waiting for SID...");
    const sid = await waitForSid();
    console.log(`SID       : ${sid}`);

    const authCrypto = encryptAuthCredentials(authConfig.username, authConfig.password, sid, authConfig.keyHex);
    const cmdauth = `C;AUTH;${sid};${authCrypto}`;

    console.log(`\n* * * Logging in as '${authConfig.username}'.`);
    console.log(`AUTH IV   : ${sid}`);
    console.log(`AUTH CRYPT: ${authCrypto}`);
    await sendAndWait(cmdauth, 'AUTH');

    console.log("\n* * * Sending NOOP command.");
    console.log(cmdnoop);
    await sendAndWait(cmdnoop, 'NOOP');

    console.log("\n* * * Open communication channel.");
    console.log(cmdopen);
    await sendAndWait(cmdopen, 'OPEN');

    console.log("\n* * * Set filter.");
    console.log(cmdFilter);
    await sendAndWait(cmdFilter, 'SF');

    console.log("\n* * * Send event.");
    console.log(sendEvent);
    ws.send(sendEvent);
  }
  catch (error) {
    console.error(`Command sequence failed: ${error.message}`);
    ws.close();
    return;
  }

  console.log("---------------------------------------------------------");
  console.log("Waiting for events (Abort with ctrl+c)")

  console.log("Will only receive CLASS1.CONTROL, TurnOn now after filter");
  console.log("is set");
  console.log("---------------------------------------------------------");

});
  
ws.on('message', function incoming(data) {
  const message = toTextMessage(data);
  const parsed = parseWs1Message(message);

  console.log("\n* * * Raw message received:", message);

  if (parsed.kind === 'positive') {
    console.log("\n* * * Positive response");
    console.log(`Command   : ${parsed.command}`);
    if (parsed.details.length) {
      console.log(`Details   : ${parsed.details.join(' ; ')}`);
    }
  }
  else if (parsed.kind === 'negative') {
    console.log("\n* * * Negative response");
    console.log(`Command   : ${parsed.command}`);
    console.log(`ErrorCode : ${parsed.errorCode}`);
    console.log(`ErrorText : ${parsed.errorText}`);
  }
  else if (parsed.kind === 'event') {
    console.log("\n* * * Event received");
    if (parsed.event) {
      console.log(`Class/Type: ${parsed.event.vscpClass}/${parsed.event.vscpType}`);
      console.log(`Head/ObId : ${parsed.event.head}/${parsed.event.obid}`);
      console.log(`Date/Time : ${parsed.event.datetime}`);
      console.log(`Timestamp : ${parsed.event.timestamp}`);
      console.log(`GUID      : ${parsed.event.guid}`);
      console.log(`Data      : [${parsed.event.data.join(', ')}]`);
    }
    else {
      console.log(`Event raw : ${parsed.raw}`);
    }
  }
  else {
    console.log("\n* * * Message received");
    console.log(parsed.raw);
  }

  if (pendingSidResolver) {
    const sid = extractSid(parsed);
    if (sid) {
      pendingSidResolver.resolve(sid);
      return;
    }

    if (parsed.kind === 'negative' && (parsed.command || '').toUpperCase() === 'CHALLENGE') {
      pendingSidResolver.reject(
        new Error(`Failed to get SID: ${parsed.errorCode} ${parsed.errorText}`)
      );
      return;
    }
  }

  if (pendingRequest && (parsed.kind === 'positive' || parsed.kind === 'negative')) {
    if (!matchesExpectedCommand(parsed, pendingRequest.expected)) {
      return;
    }

    if (parsed.kind === 'positive') {
      pendingRequest.resolve(parsed);
    }
    else {
      pendingRequest.reject(
        new Error(`Negative response for ${parsed.command}: ${parsed.errorCode} ${parsed.errorText}`)
      );
    }
  }
});
